strawberryfields/api/result.py

```diff
@@ -107,6 +103,70 @@ def state(self):
             raise AttributeError("The state is undefined for a stateless computation.")
         return self._state
 
+    @staticmethod
+    def combine_samples(samples_list, mode_order):
+        """Combine samples measured at different times into one nested array.
+
+        Args:
+            samples_list (list[array, tensor]): the sample measurements in a list
+            mode_order (list[array]): The nested mode order for the measured modes. Must have the
+                same structure as samples_list.
+
+        Returns:
+            array: the samples in ascending mode order with shape ``(shots, measured_modes)``
+        """
+
+        shots = len(samples_list[0])
+
+        mode_order_flat = np.hstack(mode_order)
+        len_modes = len(set(mode_order_flat))
+
+        # remove duplicate mode-measurements and only return the modes from the last measurement
+        if len(mode_order_flat) != len_modes:
+            # change to list to allow updates via reference, and split samples into shots
+            mode_order_list = [list(m) for m in mode_order]
+            samples_shot_list = [[list(s[i]) for s in samples_list] for i in range(shots)]
+
+            combos = it.combinations(range(len(mode_order_list)), 2)
+
+            # go through each combination of mode orders (one per measurement) and delete
+            # the first occurence of any duplicates; also delete corresponding sample(s)
+            for a, b in it.combinations(mode_order_list, 2):
+                intersect, *idx = np.intersect1d(a, b, return_indices=True)
+                k = combos.__next__()[0]
+                for i, inter in enumerate(intersect):
+                    a.remove(inter)
+                    for j in range(shots):
+                        # remove last sample in shot array so it doesn't disturb the indices
+                        s_val = samples_shot_list[j][k][idx[k][len(intersect) - i - 1]]
+                        samples_shot_list[j][k].remove(s_val)
+
+            # reshape into a list of arrays, ignoring empty ones
+            mode_order = [np.array(m) for m in mode_order_list if m]
+            samples_list = [np.array([s]) for s in zip(*samples_shot_list) if s[0]]
+
+            shapes = np.cumsum([len(m) for m in mode_order])
+
+        # create the correct mode order if not all modes are measured
+        mode_order_flat = np.hstack(mode_order)
+        if np.max(mode_order_flat) + 1 != len_modes:
```

Comment:
(Minor suggestion): could help readability
```suggestion
        not_all_modes_measured = np.max(mode_order_flat) + 1 != len_modes
        if not_all_modes_measured:
```